File: Q30.cpp
//Q30.cpp
class Solution {
public:
    vector<int> findSubstring(string s, vector<string>& w) {
        int n=s.size(), m=w.size(), L=w[0].size(), tot=m*L;
        unordered_map<string,int> need;
        for(auto &x:w) need[x]++;
        vector<int> res;
        for(int off=0; off<L; off++){
            int cnt=0; unordered_map<string,int> win;
            for(int l=off, r=off; r+L<=n; r+=L){
                string t=s.substr(r,L);
                if(need.count(t)){
                    win[t]++; cnt++;
                    while(win[t]>need[t]){
                        win[s.substr(l,L)]--; l+=L; cnt--;
                    }
                    if(cnt==m){
                        res.push_back(l);
                        win[s.substr(l,L)]--; l+=L; cnt--;
                    }
                } else {
                    win.clear(); cnt=0; l=r+L;
                }
            }
        }
        return res;
    }
};

File: Q30.py
#Q30.py
class Solution:
    def findSubstring(self, s, w):
        n=len(s); m=len(w); L=len(w[0]); tot=m*L
        need={}
        for x in w: need[x]=need.get(x,0)+1
        res=[]
        for off in range(L):
            win={}; cnt=0; l=off
            for r in range(off, n-L+1, L):
                t=s[r:r+L]
                if t in need:
                    win[t]=win.get(t,0)+1; cnt+=1
                    while win[t]>need[t]:
                        u=s[l:l+L]; win[u]-=1; l+=L; cnt-=1
                    if cnt==m:
                        res.append(l)
                        u=s[l:l+L]; win[u]-=1; l+=L; cnt-=1
                else:
                    win.clear(); cnt=0; l=r+L
        return res
